pkg/verification/verify.go

```diff
@@ -18,17 +18,175 @@ package verification
 import (
 	"bytes"
 	"crypto/x509"
-	"errors"
+	"encoding/asn1"
 	"fmt"
 	"hash"
 	"io"
+	"math/big"
 
 	"github.com/digitorus/pkcs7"
 	"github.com/digitorus/timestamp"
+	"github.com/pkg/errors"
 )
 
+type VerifyOpts struct {
+	Oid            asn1.ObjectIdentifier
+	TsaCertificate *x509.Certificate
+	Intermediates  []*x509.Certificate
+	Roots          []*x509.Certificate
+	Nonce          *big.Int
+	Subject        string
+	HashAlgorithm  hash.Hash
+	HashedMessage  []byte
+}
+
+// Verify the TSR's certificate identifier matches a provided TSA certificate
+func verifyESSCertID(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TsaCertificate == nil {
+		return nil
+	}
+
+	errMessage := ""
+
+	if !bytes.Equal(opts.TsaCertificate.RawIssuer, tsaCert.RawIssuer) {
+		errMessage += "TSR cert issuer does not match provided TSA cert issuer"
+	}
+
+	if opts.TsaCertificate.SerialNumber.Cmp(tsaCert.SerialNumber) != 0 {
+		if errMessage != "" {
+			errMessage += ", TSR cert issuer does not match provided TSA cert issuer"
+		} else {
+			errMessage = "TSR cert issuer does not match provided TSA cert issuer"
+		}
+	}
+
+	if errMessage != "" {
+		return errors.New(errMessage)
+	}
+	return nil
+}
+
+// Verify the leaf certificate's subject and/or subject alternative name matches a provided subject
+func verifyLeafCertSubject(cert *x509.Certificate, opts VerifyOpts) error {
+	if opts.Subject == "" {
+		return nil
+	}
+	
+	leafCertSubject := cert.Subject.String()
+	if leafCertSubject != opts.Subject {
+		return fmt.Errorf("Leaf cert subject %s does not match provided subject %s", leafCertSubject, opts.Subject)
+	}
+	return nil
+}
+
+// If embedded in the TSR, verify the TSR's leaf certificate matches a provided TSA certificate
+func verifyEmbeddedLeafCert(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TsaCertificate != nil && !opts.TsaCertificate.Equal(tsaCert) {
+		return fmt.Errorf("certificate embedded in the TSR does not match the provided TSA certificate")
+	}
+	return nil
+}
+
+func verifyLeafCert(ts timestamp.Timestamp, opts VerifyOpts) error {
+	if len(ts.Certificates) == 0 && opts.TsaCertificate == nil {
+		return fmt.Errorf("A leaf certificate must be present the in TSR or as a verify command argument")
+	}
+
+	errMsg := "failed to verify leaf cert"
+
+	var leafCert *x509.Certificate
+	if len(ts.Certificates) != 0 {
+		leafCert = ts.Certificates[0]
+		if opts.TsaCertificate != nil && !leafCert.Equal(opts.TsaCertificate) {
+			return fmt.Errorf("The leaf certificate included in the TSR does not match the one provied as a verify command argument")
```

Comment:
```suggestion
			return fmt.Errorf("leaf certificate included in the TSR does not match the one provided as a verify option")
```

Reply:
nit elsewhere, error messages should start lower-cased (i don't love this as go style, but oh well haha)